### PR TITLE
fix(table):table grouping columns does not apply sorting

### LINKED ISSUE #50084: Sorting on children column doesn't work after updating to 5.19.3
### Reproduction link

[Edit on StackBlitz](https://stackblitz.com/edit/antd-reproduce-5x-fx5tsy?file=demo.tsx)

### Steps to reproduce

I have attached a working sample running on antd 5.19.2. To reproduce update to 5.19.3 and then sorting will not work.

### What is expected?

Sorting should work.

### What is actually happening?

Sorting is not working

| Environment | Info |
| --- | --- |
| antd | 5.19.3 |
| React | 18.3.1 |
| System | Windows 10 |
| Browser | Google Chrome 126.0.6478.183 |

### LINKED ISSUE #50154: Ant Design Table grouping columns does not apply sorting (5.19.4)
### Reproduction link

[https://ant.design/~demos/table-demo-grouping-columns](https://ant.design/~demos/table-demo-grouping-columns)

### Steps to reproduce

Either click on Age header or sorting arrows for this grouping column child

### What is expected?

expected to sort Column Age when clicking on Age header

### What is actually happening?

nothing occurs

| Environment | Info |
| --- | --- |
| antd | 5.19.4 |
| React | 18.3.1 |
| System | MacOS 14.0 |
| Browser | Google Chrome Version 127.0.6533.72 (Official Build) (arm64) |

### PR DESCRIPTION
**Pull request by Zyf665**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
ref https://github.com/ant-design/ant-design/pull/49533
fix https://github.com/ant-design/ant-design/issues/50154
fix https://github.com/ant-design/ant-design/issues/50084

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   fixed table grouping columns does not apply sorting       |
| 🇨🇳 Chinese |   修复表格中列存在分组时，排序不生效的问题        |

### CODE REVIEW

**Comment by li-jia-nan** [comment]
这个我记得都修过好几次了

File: components/table/hooks/useSorter.tsx [hunk 1]
```diff
@@ -394,15 +394,22 @@ export default function useFilterSorter<RecordType>({
     collectSortStates(mergedColumns, true),
   );
 
+  const getColumnKeys = (columns: ColumnsType<RecordType>, keys: Key[] = []): Key[] => {
+    columns.forEach((item, index) => {
+      keys.push(getColumnKey(item, getColumnPos(index)));
```
Comment (li-jia-nan) [inline components/table/hooks/useSorter.tsx:399]: 直接改原始值不太好吧，是不是应该 clone 一个新数据出来再操作

Comment (Zyf665) [reply components/table/hooks/useSorter.tsx:399]: 这块是把列里面的key提取出来了   改原始值指的是什么

Comment (li-jia-nan) [reply components/table/hooks/useSorter.tsx:399]: keys.push 这里，看到你直接修改了参数

Comment (Zyf665) [reply components/table/hooks/useSorter.tsx:399]: ok,了解

**Comment by wanpan11** [comment]
Changelog 记得补充下 😶‍🌫️

**Comment by afc163** [comment]
issue 关联的是错的。

**Comment by afc163** [comment]
应该关联 https://github.com/ant-design/ant-design/issues/50084 这个 issue。

**Comment by afc163** [comment]
Changelog 请认真写下。

**Review by afc163 — APPROVED** [review]

File: components/table/hooks/useSorter.tsx [hunk 2]
```diff
@@ -394,16 +394,25 @@ export default function useFilterSorter<RecordType>({
     collectSortStates(mergedColumns, true),
   );
 
+  const getColumnKeys = (columns: ColumnsType<RecordType>, pos?: string): Key[] => {
```
Comment (crazyair) [inline components/table/hooks/useSorter.tsx:397]: 如果这里用 columns?:xxx 了，下面就不需要 as 了

**Pull request merged**